utilities/docs/json-schema-generate.8
.\" Manpage for json-schema-generate.
.\" Documentation is under the same licence as the Walbottle package.
.TH man 8 "29 Sep 2015" "1.0" "json-schema-generate man page"

.SH NAME
.IX Header "NAME"
json-schema-generate — JSON instance generator

.SH SYNOPSIS
.IX Header "SYNOPSIS"
\fBjson-schema-generate \fPschema-file\fB [\fPschema-file\fB …] [-q] [-v] [-n]
[-j] [-f \fPformat-name\fB] [--c-variable-name \fPvariable_name\fB]
[--show-timings]

.SH DESCRIPTION
.IX Header "DESCRIPTION"
\fBjson-schema-generate\fP is a utility for generating JSON instances which may
or may not be valid for a given JSON schema. Both valid and invalid instances
are designed to test corner cases of the parser which implements the given JSON
schema.

See the documentation for libwalbottle for instructions on how to integrate
json-schema-generate into a typical test suite build process.

Valid instances (well-formed JSON which validates against the schema),
invalid instances (well-formed JSON which does not validate against the schema)
and invalid JSON (non-well-formed JSON) are generated by this utility by
default. This can be controlled using the \fB--valid-only\fP,
\fB--invalid-only\fP and \fB--no-invalid-json\fP options.

Instances are printed on standard output. By default, one JSON instance is
outputted per line. A C array of escaped JSON strings may be outputted instead
by choosing \fB--format=c\fP. This will be in the format below — the
\fB--c-variable-name\fP option may be used to change the variable name from its
default:

.RS
/* Generated by json-schema-generate. Do not modify. */

#include <stddef.h>

static const struct { const char *json; size_t size; unsigned int is_valid; } json_instances[] = {
.br
	{ "\\"some JSON instances\\"", 21, 0 },  /* 0 */
.br
	…
.br
};
.RE

The output formats are guaranteed to not change without a change in Walbottle’s
major version number. This includes the C structure and types.

For information on JSON Schema, see \fIhttp://json-schema.org/\fP.

.SH OPTIONS
.IX Header "OPTIONS"
.IP "\fB\-q \-\-quiet\fP"
Do not print progress output during generation.
.IP "\fB\-v \-\-valid\-only\fP"
Only output valid instances (well-formed JSON which validates against the
schema).
.IP "\fB\-n \-\-invalid\-only\fP"
Only output invalid instances (well-formed JSON which does not validate against
the schema).
.IP "\fB\-j \-\-no\-invalid\-json\fP"
Do not output invalid JSON (non-well-formed JSON).
.IP "\fB\-f \-\-format\fP [plain|c]"
Output in plain format (one JSON instance per line) or C format (all JSON
instances escaped as strings in an array).
.IP "\fB\-\-c\-variable\-name\fP variable_name"
Set the name of the JSON instance array output if using \-\-format=c. This
option must only be used with C-format output.
.IP "\fB\-\-show\-timings\fP"
Print debugging and timing information for all schemas and sub-schemas after
printing the generated instances. This is intended to be used as guidance for
optimising JSON schema files. If multiple schema files are provided, timing
information is printed for each in turn.

.SH "ENVIRONMENT"
.IX Header "ENVIRONMENT"
json-schema-generate supports the standard GLib environment variables for
debugging. These variables are \fBnot\fP intended to be used in production:
.IP \fBG_MESSAGES_DEBUG\fR 4
.IX Item "G_MESSAGES_DEBUG"
This variable can contain one or more debug domain names to display debug output
for. The value \fIall\fP will enable all debug output. The default is for no
debug output to be enabled.

.SH "EXIT STATUS"
.IX Header "EXIT STATUS"
json-schema-generate may return one of several error codes if it encounters
problems.

.IP "0" 4
No problems occurred. Instances were generated successfully for the given
schemas.
.IP "1" 4
.IX Item "1"
An invalid option was passed to json-schema-generate on startup.
.IP "2" 4
.IX Item "2"
One of the JSON schemas was not well-formed or did not validate against the
meta-schema.

.SH EXAMPLES
.IX Header "EXAMPLES"
Here is an example of generating JSON instances for a given schema, in C format:
.br
.PP
\fBjson-schema-generate\fP --format=c --c-variable-name=my_schema_json_instances
/path/to/my-schema.schema.json

.SH "SEE ALSO"
.IX Header "SEE ALSO"
.I json-validate(8)
.I json-schema-validate(8)

.SH BUGS
.IX Header "BUGS"
Any bugs which are found should be reported on the project website:
.br
.I https://gitlab.com/walbottle/walbottle

.SH AUTHOR
.IX Header "AUTHOR"
Collabora Ltd.

.SH COPYRIGHT
.IX Header "COPYRIGHT"
Copyright © 2015 Collabora Ltd.
.PP


NEWS
Overview of changes from Walbottle 0.2.0 to Walbottle 0.3.0
===========================================================

Major changes:

API changes:

Bugs fixed:

Updated translations:


Overview of changes from Walbottle 0.1.0 to Walbottle 0.2.0
===========================================================

Major changes:
 • Improved support for the ‘default’ keyword
 • Support multiple output formats in json-schema-generate
 • Move from Gitorious to GitLab
 • Performance improvements
 • Improved support for property keywords (‘minProperties’, etc.)
 • Improved support for item keywords (‘additionalItems’, etc.)
 • Improved support for array keywords (‘oneOf’, etc.)
 • Install versioned copies of utility programs
   (i.e. json-schema-generate-0, symlinked from json-schema-generate)
 • Commit to stability of output formats from json-schema-generate
 • Add man pages

API changes:
 • Add wbl_generated_instance_is_valid()
 • Rename WblSchema.generate_instances to WblSchema.generate_instance_nodes


Initial release of Walbottle 0.1.0
==================================

Major changes:
 • Initial version of the project
 • Schema validation, instance validation and instance generation support
 • Nearly full implementation of the json-schema-core and json-schema-validation
   specifications
 • Not API stable


utilities/json-schema-generate.c
/* -*- Mode: C; indent-tabs-mode: t; c-basic-offset: 8; tab-width: 8 -*- */
/*
 * Walbottle
 * Copyright (C) Philip Withnall 2015, 2016 <philip@tecnocode.co.uk>
 *
 * Walbottle is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * Walbottle is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with Walbottle.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "config.h"

#include <locale.h>
#include <glib.h>
#include <glib/gi18n.h>
#include <stdio.h>

#include "wbl-schema.h"
#include "wbl-meta-schema.h"
#include "utilities/wbl-utilities.h"

/* Exit statuses. */
typedef enum {
	/* Success. */
	EXIT_SUCCESS = 0,
	/* Error parsing command line options. */
	EXIT_INVALID_OPTIONS = 1,
	/* JSON schema could not be parsed. */
	EXIT_INVALID_SCHEMA = 2,
} ExitStatus;

/* Output formats. */
typedef enum {
	FORMAT_PLAIN = 0,
	FORMAT_C,
} OutputFormat;

/* Indexed by OutputFormat. */
static const gchar *output_formats[] = {
	"plain",
	"c",
};

G_STATIC_ASSERT (G_N_ELEMENTS (output_formats) == FORMAT_C + 1);

static gint
sort_schema_info_cb (gconstpointer a,
                     gconstpointer b)
{
	WblSchemaInfo *info_a = *((WblSchemaInfo **) a);
	WblSchemaInfo *info_b = *((WblSchemaInfo **) b);
	gint64 time_a, time_b;

	time_a = wbl_schema_info_get_generation_time (info_a);
	time_b = wbl_schema_info_get_generation_time (info_b);

	return time_b - time_a;
}

/* Command line parameters. */
static gboolean option_quiet = FALSE;
static gboolean option_valid_only = FALSE;
static gboolean option_invalid_only = FALSE;
static gboolean option_no_invalid_json = FALSE;
static gchar *option_format = NULL;
static gchar **option_schema_filenames = NULL;
static gchar *option_c_variable_name = NULL;
static gboolean option_show_timings = FALSE;

static const GOptionEntry entries[] = {
	{ "quiet", 'q', 0, G_OPTION_ARG_NONE, &option_quiet,
	  N_("Quieten command line output"), NULL },
	{ "valid-only", 'v', 0, G_OPTION_ARG_NONE, &option_valid_only,
	  N_("Only output valid JSON instances"), NULL },
	{ "invalid-only", 'n', 0, G_OPTION_ARG_NONE, &option_invalid_only,
	  N_("Only output invalid JSON instances"), NULL },
	{ "no-invalid-json", 'j', 0, G_OPTION_ARG_NONE,
	  &option_no_invalid_json,
	  N_("Disable generation of invalid JSON vectors"), NULL },
	{ "format", 'f', 0, G_OPTION_ARG_STRING, &option_format,
	  N_("Output format (‘plain’ [default], ‘c’)"), NULL },
	{ "c-variable-name", 0, 0, G_OPTION_ARG_STRING, &option_c_variable_name,
	  N_("Vector array variable name (only with --format=c; default "
	     "‘json_instances’"), NULL },
	{ "show-timings", 0, 0, G_OPTION_ARG_NONE, &option_show_timings,
	  N_("Print timing information to stderr after outputting generated "
	     "instances"), NULL },
	{ G_OPTION_REMAINING, 0, 0, G_OPTION_ARG_FILENAME_ARRAY,
	  &option_schema_filenames,
	  N_("JSON schema files to generate from"),
	  N_("JSON-SCHEMA [JSON-SCHEMA …]") },
	{ NULL, },
};

int
main (int argc, char *argv[])
{
	GOptionContext *context = NULL;  /* owned */
	ExitStatus retval = EXIT_SUCCESS;
	guint i;
	GPtrArray/*<owned WblSchema>*/ *schemas = NULL;  /* owned */
	WblGenerateInstanceFlags flags;
	OutputFormat output_format;
	gboolean output_format_set = FALSE;
	GError *error = NULL;
	gboolean generated_any_valid_instances = FALSE;
	gboolean generated_any_invalid_instances = FALSE;
	gboolean use_colour_stderr;
	const gchar *bold_escape, *reset_escape;

#if !GLIB_CHECK_VERSION (2, 35, 0)
	g_type_init ();
#endif

	setlocale (LC_ALL, "");

	/* Can we use colour output? */
	use_colour_stderr = wbl_is_colour_supported (stderr);

	/* Redirect debug output to stderr so that stdout is purely generated
	 * test vectors. */
	g_log_set_default_handler (wbl_log, NULL);

	/* Command line parsing. */
	context = g_option_context_new (_("— generate test vectors from JSON "
	                                  "schemas"));
	g_option_context_set_summary (context,
	                              _("Generate valid and invalid instances "
	                                "of one or more JSON Schemas. These "
	                                "can be used as test vectors in unit "
	                                "tests for code which parses documents "
	                                "which should conform to all of these "
	                                "schemas. Schemas are outputted one "
	                                "per line.\n\nRead about JSON Schema "
	                                "here: http://json-schema.org/."));
	g_option_context_add_main_entries (context, entries, PACKAGE_NAME);

	if (!g_option_context_parse (context, &argc, &argv, &error)) {
		gchar *message;

		message = g_strdup_printf (_("Option parsing failed: %s"),
		                           error->message);
		g_printerr ("%s: %s\n", argv[0], message);
		g_free (message);

		g_clear_error (&error);

		retval = EXIT_INVALID_OPTIONS;
		goto done;
	}

	/* Validate the options. */
	if (option_format == NULL) {
		/* Default. */
		output_format = FORMAT_PLAIN;
		output_format_set = TRUE;
	}

	for (i = 0;
	     i < G_N_ELEMENTS (output_formats) && !output_format_set; i++) {
		if (g_strcmp0 (output_formats[i], option_format) == 0) {
			output_format = i;
			output_format_set = TRUE;
			break;
		}
	}

	if (!output_format_set) {
		gchar *message1, *message2;

		message1 = g_strdup_printf (_("Invalid output format ‘%s’."),
		                            option_format);
		message2 = g_strdup_printf (_("Option parsing failed: %s"),
		                            message1);
		g_printerr ("%s: %s\n", argv[0], message2);
		g_free (message2);
		g_free (message1);

		g_clear_error (&error);

		retval = EXIT_INVALID_OPTIONS;
		goto done;
	}

	if (output_format == FORMAT_C) {
		/* Don’t bother validating @option_c_variable_name; the compiler
		 * will eventually do that for us. */
		if (option_c_variable_name == NULL ||
		    g_strcmp0 (option_c_variable_name, "") == 0) {
			g_free (option_c_variable_name);
			option_c_variable_name = g_strdup ("json_instances");
		}
	} else if (option_c_variable_name != NULL) {
		const gchar *message = NULL;

		message = _("Option --c-variable-name may only be specified "
		            "with --format=c.");
		g_printerr ("%s: %s\n", argv[0], message);

		retval = EXIT_INVALID_OPTIONS;
		goto done;
	}

	/* Load the schemas. */
	schemas = g_ptr_array_new_with_free_func ((GDestroyNotify) g_object_unref);

	for (i = 0;
	     option_schema_filenames != NULL && option_schema_filenames[i] != NULL;
	     i++) {
		WblSchema *schema = NULL;  /* owned */

		schema = wbl_schema_new ();
		wbl_schema_load_from_file (schema, option_schema_filenames[i],
		                           &error);

		if (error != NULL) {
			if (!option_quiet) {
				gchar *message;

				message = g_strdup_printf (_("Invalid JSON schema ‘%s’: %s"),
				                           option_schema_filenames[i],
				                           error->message);
				g_printerr ("%s: %s\n", argv[0], message);
				g_free (message);
			}

			g_clear_error (&error);
			g_clear_object (&schema);

			retval = EXIT_INVALID_SCHEMA;
			goto done;
		}

		g_ptr_array_add (schemas, schema);  /* transfer */
	}

	/* Set up the generation flags. */
	flags = WBL_GENERATE_INSTANCE_NONE;

	if (option_valid_only) {
		flags |= WBL_GENERATE_INSTANCE_IGNORE_INVALID;
	}
	if (option_invalid_only) {
		flags |= WBL_GENERATE_INSTANCE_IGNORE_VALID;
	}
	if (!option_no_invalid_json) {
		flags |= WBL_GENERATE_INSTANCE_INVALID_JSON;
	}

	/* Initial output. This format is part of the json-schema-generate ABI
	 * and cannot be modified without a major version break. */
	if (output_format == FORMAT_C) {
		g_print ("/* Generated by %s. Do not modify. */\n\n", argv[0]);
		g_print ("#include <stddef.h>\n\n");
		g_print ("static const struct { const char *json; size_t size; unsigned int is_valid; } %s[] = {\n",
		         option_c_variable_name);
	}

	/* Generate from each of the schemas. */
	for (i = 0; i < schemas->len; i++) {
		WblSchema *schema;  /* unowned */
		GPtrArray/*<owned WblGeneratedInstance>*/ *instances = NULL;  /* owned */
		guint j;

		schema = schemas->pdata[i];
		instances = wbl_schema_generate_instances (schema, flags);

		/* Print out the instances. This format is part of the
		 * json-schema-generate ABI and cannot be modified without a
		 * major version break. */
		for (j = 0; j < instances->len; j++) {
			WblGeneratedInstance *instance;  /* unowned */
			const gchar *json;
			gboolean is_valid;

			instance = instances->pdata[j];
			json = wbl_generated_instance_get_json (instance);
			is_valid = wbl_generated_instance_is_valid (instance);
			generated_any_valid_instances |= is_valid;
			generated_any_invalid_instances |= !is_valid;

			switch (output_format) {
			case FORMAT_PLAIN:
				g_print ("%s\n", json);
				break;
			case FORMAT_C: {
				gchar *escaped = g_strescape (json, "");
				g_print ("\t{ \"%s\", %" G_GSIZE_FORMAT ", "
				         "%u },  /* %u */\n",
				         escaped, strlen (json),
				         is_valid ? 1 : 0, j);
				g_free (escaped);
				break;
			}
			default:
				g_assert_not_reached ();
			}
		}

		g_ptr_array_unref (instances);
	}

	/* Final output. */
	if (output_format == FORMAT_C) {
		g_print ("};\n");
	}

	/* Timing output for each of the schemas. */
	if (use_colour_stderr) {
		/* See: http://misc.flogisoft.com/bash/tip_colors_and_formatting */
		bold_escape = "\033[1m";
		reset_escape = "\033[0m";
	} else {
		bold_escape = "";
		reset_escape = "";
	}

	if (option_show_timings) {
		for (i = 0; i < schemas->len; i++) {
			WblSchema *schema;  /* unowned */
			GPtrArray/*<owned WblSchemaInfo>*/ *infos = NULL;  /* owned */
			guint j;

			schema = schemas->pdata[i];
			infos = wbl_schema_get_schema_info (schema);

			/* Sort the schema info by the generation time. */
			g_ptr_array_sort (infos, sort_schema_info_cb);

			/* Print timing information. */
			g_printerr ("%s%s%s timings:\n",
			            bold_escape, option_schema_filenames[i],
			            reset_escape);

			for (j = 0; j < infos->len; j++) {
				WblSchemaInfo *info = infos->pdata[j];
				guint n_instances;
				gint64 generation_time;
				gdouble time_per_instance;

				n_instances = wbl_schema_info_get_n_instances_generated (info);
				generation_time = wbl_schema_info_get_generation_time (info);

				if (n_instances > 0)
					time_per_instance = (gdouble) generation_time / n_instances;
				else
					time_per_instance = 0.0;

				g_printerr (" • %s%u%s generation took %"
				            G_GINT64_FORMAT "μs, %u times, "
				            "generating %u instances "
				            "(%.2fμs⋅instance⁻¹)\n",
				            bold_escape,
				            wbl_schema_info_get_id (info),
				            reset_escape,
				            generation_time,
				            wbl_schema_info_get_n_times_generated (info),
				            n_instances,
				            time_per_instance);
			}

			g_printerr ("%s%s%s schemas (total: %u):\n",
			            bold_escape, option_schema_filenames[i],
			            reset_escape, infos->len);

			for (j = 0; j < infos->len; j++) {
				WblSchemaInfo *info = infos->pdata[j];
				gchar *json = NULL;

				json = wbl_schema_info_build_json (info);
				g_printerr (" • %s%u%s:\n"
				            "      %s\n",
				            bold_escape,
				            wbl_schema_info_get_id (info),
				            reset_escape,
				            json);
				g_free (json);
			}

			g_ptr_array_unref (infos);
		}
	}

	/* Sanity check. */
	if (!option_invalid_only && !generated_any_valid_instances) {
		g_printerr ("%s: Warning: Failed to generate any valid "
		            "instances. Test coverage may be low. This may "
		            "indicate a bug in Walbottle; please report it.\n",
		            argv[0]);
	} else if (!option_valid_only && !generated_any_invalid_instances) {
		g_printerr ("%s: Warning: Failed to generate any invalid "
		            "instances. Test coverage may be low. This may "
		            "indicate a bug in Walbottle; please report it.\n",
		            argv[0]);
	}

done:
	if (context != NULL) {
		g_option_context_free (context);
	}
	if (schemas != NULL) {
		g_ptr_array_unref (schemas);
	}

	return retval;
}
